nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-avro-record-utils/src/main/java/org/apache/nifi/avro/AvroTypeUtil.java

```diff
@@ -672,10 +673,20 @@ private static Object convertToAvroObject(final Object rawValue, final Schema fi
             case RECORD:
                 final GenericData.Record avroRecord = new GenericData.Record(fieldSchema);
 
-                final Record record = (Record) rawValue;
-                for (final RecordField recordField : record.getSchema().getFields()) {
-                    final Object recordFieldValue = record.getValue(recordField);
-                    final String recordFieldName = recordField.getFieldName();
+                final Set<Map.Entry<String, Object>> entries;
+                if (rawValue instanceof Map) {
+                    final Map<String, Object> map = (Map<String, Object>) rawValue;
+                    entries = map.entrySet();
+                } else if (rawValue instanceof Record) {
+                    entries = new HashSet<>();
+                    final Record record = (Record) rawValue;
+                    record.getSchema().getFields().forEach(field -> entries.add(new AbstractMap.SimpleEntry<>(field.getFieldName(), record.getValue(field))));
```

Comment:
I would recommend adding a check for a `RECORD` data type in each field and then doing a recursive call to get a `Map<String, Object`> for that. Otherwise you're going to end up with only a partial conversion which could potentially introduce some ugly issues when trying to convert a complicated record structure into a JSON Map<String, Object> like structure.

Reply:
The recursive conversion is already being taken care of later on, at line 696: `final Object converted = convertToAvroObject(recordFieldValue, field.schema(), fieldName + "/" + recordFieldName, charset);`, where each field's value (pulled via `record.getValue(field)` earlier for a RECORD, or directly pulled out from the MAP) is passed recursively to `convertToAvroObject`, and the result of that is put into the Avro record.

Reply:
We had some issues in the past, but I did some additional experimentation with JUnit and found that it finally appears to be OBE now.